### DISCUSSION TITLE
Future Plans for Alpaca

### DISCUSSION THREAD

**Discussion by Jeffser** [Jeffser/Alpaca#248] [Ideas]
Hi guys, I'm almost finished with this semester in college, I wanted to share with you guys the plans I have for the two week vacation I get at the end of this month!

## General Rewrite of the App

So you guys know how I said that I made Alpaca just as a way of learning GTK? soooo I didn't really know what I was doing when I first started developing the app, there are no classes and everything is held together with tape. So I wanted to do it properly with multiple files and custom widgets and all of that good stuff.

## Web Plugins

I've been toying with this idea for a while now, I wanted to add the option for the user to write a JSON file with HTML IDs and Classes so that a website extractor can grab the right text from a website to parse to the models, something like: title=#TITLE_H1, body=DIV.body.small or something like that, I think it can be a cool way of integrating web browsing with the app, of course I would write a lot of them for the more popular websites and will always leave the current web extractor thingy as a fallback.

---

So that's it, I may or may not do everything in those 2 weeks, it depends mostly on how much my friends want to hang out.

**Comment by Jeffser** [comment]
I really need to rewrite the app tho, it's getting out of hand

**Reply by olumolu** [reply]
If you really think to rewrite the app i think switching to a language like rust maybe better. But still if you are comfortable with python there is no issue with it.

**Reply by loulou64490** [reply]
there is a dedicated Ollama library for Python developed by the team, so I think it's really really more comfortable

**Reply by olumolu** [reply]
Python is too bad for performance
Go can be a good option easy syntax.
But if possible a low level language like rust can be better and it will give many things along with memory safety and more robust application.

**Reply by Jeffser** [reply]
I told you before, I'm not changing the language, I'm confortable with Python and making the app faster doesn't matter at all since Ollama will slow it down regardless

**Reply by olumolu** [reply]
Have you started the rewrite

**Reply by Jeffser** [reply]
Nope, my vacation starts at the end of the month

**Reply by olumolu** [reply]
What part of it you think you need to rewrite 
Or you want to rewrite this from scratch

**Comment by Jeffser** [comment]
I got some free time tonight and decided to rewrite the message element, now it is possible to use all the formatting stuff with all the user messages (code blocks, markup, tables)

![image](https://github.com/user-attachments/assets/71c9252d-ee8b-43b7-b6b1-2957ef8f33f7)

**Reply by Jeffser** [reply]
![image](https://github.com/user-attachments/assets/6f0d4bed-d986-43e0-b935-2d44e57a412f)

I made my first proper custom widget!

**Reply by Jeffser** [reply]
alright, the message widgets are ready, I just need to add the edit, delete, regenerate and copy functions

**Comment by Jeffser** [comment]
[Screencast from 2024-08-24 19-41-01.webm](https://github.com/user-attachments/assets/dd7fa720-8342-43d9-9bce-9fa5b53558b3)

I rewrote the chat window widget, it still not ready but I switched to Gtk.Stack, that way I can preload every chat and it has nice animations

**Comment by Jeffser** [comment]
Probably the greatest upgrade to the app

(the first message always takes the longest to be sent because it needs to start on the model)

[Screencast from 2024-08-25 19-04-32.webm](https://github.com/user-attachments/assets/ab5003b9-f582-4823-baab-95b30c022b81)

**Reply by Jeffser** [reply]
Basically having multiple chats at the same time! you can also manage models, create, export and import chats and whatever else you might want to do whilst generating a response!

**Reply by Jeffser** [reply]
If you delete the chat it stops the message of course

**Reply by olumolu** [reply]
> Probably the greatest upgrade to the app
> 
> (the first message always takes the longest to be sent because it needs to start on the model)
> Screencast.from.2024-08-25.19-04-32.webm

Why not show model loading when the model is being loaded.

**Comment by Jeffser** [comment]
Welp that's something

![image](https://github.com/user-attachments/assets/ea6f38da-e803-41f2-a3ac-c38992eb1df1)

**Comment by Jeffser** [comment]
Now this is what I call a great use of custom widgets, they look so cool and organized on the GTK inspector thingy :D

![image](https://github.com/user-attachments/assets/5a548753-8b51-410c-810d-85c0a04e6afe)

**Comment by Jeffser** [comment]
New widget for models being pulled

![image](https://github.com/user-attachments/assets/4e4526f4-0e84-4c53-abfd-6ee24f16f8e6)

**Reply by Jeffser** [reply]
Soooo it turns out Ollama can't actually stop a model download, what I was doing was interrupting the stream of data from Ollama to Alpaca, that usually works but apparently not for downloads, once the user restarts Alpaca the download restarts in the background, long story short, no more stop button :)

![image](https://github.com/user-attachments/assets/9591ab49-3ba3-45da-9970-2e7f761c82fe)

**Reply by olumolu** [reply]
But there should be a cancellation button like if you have clicked on a big model like llama3.1 405b but you don't need that model so having a cancel button makes it easy.

**Reply by Jeffser** [reply]
it doesn't work

**Reply by Jeffser** [reply]
![image](https://github.com/user-attachments/assets/c569d05d-2bc4-4911-a787-89e36e80ac34)

Figured it out

**Comment by Jeffser** [comment]
![image](https://github.com/user-attachments/assets/1a385c6f-36c5-4ed9-aa2a-3deb3704df2f)

New widgets for local and available models

**Comment by Jeffser** [comment]
I added this warning banner

![image](https://github.com/user-attachments/assets/c1f1154f-2af7-45e6-8954-f850480eae59)

**Comment by Jeffser** [comment]
# Huge!

I moved all the loading to a different thread and sooo this is the launch time difference

## Before

![Screenshot from 2024-08-31 17-43-54](https://github.com/user-attachments/assets/a1f2dfe4-5d91-4a2e-bbc2-be625bfd888f)

## After

![Screenshot from 2024-08-31 17-45-46](https://github.com/user-attachments/assets/b222f61c-e3d0-4596-8748-91381e95433e)

So yeah, the app now launches the millisecond you click the icon, cool isn't it?

Now of course it does need to load everything, right now it looks like this when you launch it

[Screencast from 2024-08-31 17-49-11.webm](https://github.com/user-attachments/assets/0fbfe422-278f-4652-acaf-93d3f1afeacf)

That's really ugly, so I'm going to add a small dialog displaying everything that's being loaded

Edit: Fixed it!

[Screencast from 2024-08-31 18-20-04.webm](https://github.com/user-attachments/assets/c31ad84b-999d-4c03-b105-ee82c110c82c)

**Reply by loulou64490** [reply]
wow beautiful

**Reply by loulou64490** [reply]
this will be alpaca 2.0 ?

**Reply by Jeffser** [reply]
Yeah I think I made enough changes to make it a 2.0 release, I still need to do a lot of stuff before it's ready tho

**Reply by olumolu** [reply]
I was thinking if you can implement something like a timeout so for laptop we can save some battery also.
If we again use it will turn ollama on.

**Reply by olumolu** [reply]
Hey is alpaca suppoted in arm64 systems also.

**Reply by Jeffser** [reply]
yeah it is

**Reply by olumolu** [reply]
Why not release it as a beta so we can test i know you will test yourself though.

**Reply by olumolu** [reply]
Do you think about release a snap also as it is now 2.0

**Reply by loulou64490** [reply]
> Why not release it as a beta so we can test i know you will test yourself though.

Or maybe make a new branch for testing, so we can try the app before the global launch @Jeffser

**Reply by olumolu** [reply]
I have compiled. My self. You can do that

**Comment by olumolu** [comment]
Any idea how to get new
https://thisweek.gnome.org/posts/2024/06/twig-154/65678b483e333dc2a1e80a593ee9c4e7adfc84781806754585696534528.png
Pop up

**Reply by Jeffser** [reply]
That's Gnome 47, Alpaca works perfectly fine with these new dialogs but you'll have to wait until it gets released

**Comment by loulou64490** [comment]
Ho yeah I haven't seen that he push his commit

**Reply by olumolu** [reply]
What is it can you tell me in english

**Reply by loulou64490** [reply]
oops

**Reply by loulou64490** [reply]
the email answer bug

**Reply by loulou64490** [reply]
I was answering to https://github.com/Jeffser/Alpaca/discussions/248#discussioncomment-10544406

**Reply by olumolu** [reply]
Ohhh i see

**Comment by olumolu** [comment]
Switch to newer html2text what you are using is old 12 yearsold no more maintained
 I think you can switch to a different one https://github.com/Alir3z4/html2text

**Comment by olumolu** [comment]
Gnome 47 will be released within few days lets make alpaca use gnome 47 components. And what is the status for circle support?

**Reply by Jeffser** [reply]
I know, the framework isn't available on Flathub yet, I'll change it once it's available. The Gnome Circle submission is in cross-review, it might take a couple of weeks

**Reply by olumolu** [reply]
https://flathub.org/apps/org.gnome.Platform
Can you now do make it. 47
